src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java

```diff
@@ -1318,4 +1297,861 @@ public boolean isTopEntity() {
         }
     } 
     
+    @Override
+    public void accept(OWLNamedObjectVisitor visitor) {
+        readLock.lock();
+        try {
+            delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public <O> O accept(OWLNamedObjectVisitorEx<O> visitor) {
+        readLock.lock();
+        try {
+            return delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
+        writeLock.lock();
+        try {
+            delegate.setOWLOntologyManager(saveOntology);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void saveOntology() throws OWLOntologyStorageException {
+        writeLock.lock();
```

Comment:
I wouldn't mind moving it in the OWL API.

Having had a night's sleep on this, I realized most of these changes can be rewritten to delegate to methods that already acquire a lock:

@Override
public Set<OWLAxiom> getAxioms(boolean b) {
    return getAxioms(  Imports.fromBoolean(b));
}

as getAxioms(Imports) already does the required locking, the convenience/legacy variants can delegate without loss.

the saveOntology() variants can also leverage the fact that the implementation simply delegates to the manager - so if current access through the manager for saving is safe, then there is no need for read or write lock. Of course, if the manager isn't taking precautions on saving, the problem will simply show in another place.

In owlapitools we have some parallel tests that were good enough to spot a few classes of multithread bugs. I could try expanding on them to try and test this code.
If you wish to move it to OWL API, I don't have objections - we can add it as a new module. Could be this code, or equivalent code if just moving this code could cause copyright owner headaches in Stanford.